honeybadger-io/honeybadger-laravel
# Implement user informer and feedback form Blade directives

## Status
**READY**

## Description
~This PR is the first stage of work for #62 — I thought it would be better to start with adding the user informer feature first, to lay the foundation for the feedback form.~ Full thing implemented now.

### How it works
This PR adds a custom Blade directive, `@honeybadgerInformer`, that can be included in a Blade template to get the Honeybadger error ID output. It uses the styling of Laravel's default error page, but allows for the user to override the styling  and text by passing in options:


- Laravel default error page:
```php
                <div class="flex items-center pt-8 sm:justify-start sm:pt-0">
                    <div class="px-4 text-lg text-gray-500 border-r border-gray-400 tracking-wider">
                        @yield('code')
                    </div>

                    <div class="ml-4 text-lg text-gray-500 uppercase tracking-wider">
                        @yield('message')
                    </div>
                </div>
```

![image](https://user-images.githubusercontent.com/14361073/109680752-177b0a00-7b7d-11eb-8f86-2d2e8f2be778.png)

- With `@honeybadgerInformer`:

```php
                <div class="flex items-center pt-8 sm:justify-start sm:pt-0">
                    <div class="px-4 text-lg text-gray-500 border-r border-gray-400 tracking-wider">
                        @yield('code')
                    </div>

                    <div class="ml-4 text-lg text-gray-500 uppercase tracking-wider">
                        @yield('message')
                    </div>

                    @honeybadgerInformer
                </div>
```
![image](https://user-images.githubusercontent.com/14361073/109714573-d72e8280-7ba2-11eb-8f22-cbbb2094efd4.png)


- With `@honeybadgerInformer + custom styles:

```php
@honeybadgerInformer(["class" => "uppercase text-gray-500"])
```
![image](https://user-images.githubusercontent.com/14361073/109714749-14931000-7ba3-11eb-9c4c-ccd1751deba7.png)


- With `@honeybadgerInformer + custom text:
```php
@honeybadgerInformer(["text" => "HB Error"])
```

![image](https://user-images.githubusercontent.com/14361073/109714866-37bdbf80-7ba3-11eb-9c67-fc4f281dadc8.png)


### Considerations
- The name of the Blade directive. I went with `@honeybadgerInformer`, but I'm not sure that's the best. Maybe `@honeybadgerErrorId`?
- Laravel provides multiple error pages for different status codes. Howeverm to get this to work, the user has to add the directive to the base error template, not one of the status code views like `500.blade.php`. This is because the status code views inject content into specified slots in the base templates, so any content that doesn't have a specified slot in the base will be completely unstyled.
- Laravel actually has three base error templates — `minimal.blade.php` (the default), `layout.blade.php`, and `illustrated-layout,blade.php`. The user will have to add the directive to whichever templates they're using. I'll mention this in the docs.
- The directive uses the default Tailwind styling in the minimal layout, which is why we're allowing users to override the styling.
- Re the current styling: I'd have preferred the Honeybadger part to go below the "500 | Server Error", but the Laravel error page styling is pretty tightly coupled, and I couldn't figure out a way to do it without modifying the parent's styles. The good thing is that the user can place the directive wherever they want on the page and pass in the necessary classes to customise the look.
- I switched the default message to "Error ID:", as I feel "Honeybadger Error" might be somewhat more confusing to an end-user. With "Error ID", they might realise it's something they can take note of and share with support.
- I removed any need for config keys (`user_informer.enabled` and `user_informer.message`):
  - The `enabled` key is redundant: by including the Blade directive, you're explicitly enabling the informer.
  - The `message` key is useful, but I think it's better as a parameter to the Blade directive instead. The less config file items, the better.

## To test this:
- Set up a new Laravel app (see https://github.com/honeybadger-io/honeybadger-laravel/issues/67#issuecomment-779532534)
- Modify your `composer.json` so it points to your local version of honeybadger-laravel.
- Publish Laravel error pages: `php artisan vendor:publish --tag=laravel-errors`
- Switch `APP_DEBUG` in your .env file to false, otherwise you'll only see the development error page
- Add the `@honeybadgerInformer` directive in `resources/views/errors/minimal.blade.php`.
- Add a route that triggers an error, then start the app with `php artisan serve` and visit that route.
